File: content/snippets/scala2/interview_lcci_17_21_volume_of_histogram.scala
object Solution {
  def trap(height: Array[Int]): Int = {
    var left = 0; var right = height.length - 1
    var leftMax = 0; var rightMax = 0; var ans = 0
    while (left < right) {
      if (height(left) < height(right)) {
        if (height(left) >= leftMax) leftMax = height(left)
        else ans += leftMax - height(left)
        left += 1
      } else {
        if (height(right) >= rightMax) rightMax = height(right)
        else ans += rightMax - height(right)
        right -= 1
      }
    }
    ans
  }
}

File: content/snippets/go/interview_lcci_17_21_volume_of_histogram.go
package main

func trap(height []int) int {
	left, right := 0, len(height)-1
	leftMax, rightMax := 0, 0
	ans := 0
	for left < right {
		if height[left] < height[right] {
			if height[left] >= leftMax {
				leftMax = height[left]
			} else {
				ans += leftMax - height[left]
			}
			left++
		} else {
			if height[right] >= rightMax {
				rightMax = height[right]
			} else {
				ans += rightMax - height[right]
			}
			right--
		}
	}
	return ans
}
